### ISSUE TITLE
Clarify Text length requirements

### ISSUE THREAD

**Issue by DeltaF1** [indutny/VowLink#1]
In https://github.com/indutny/VowLink/blob/master/protocol.md#message there is this assertion

> Maximum text length in Text is 256 UTF-8 characters and MUST be enforced.

Does this mean that Text can contain at most 256 Unicode codepoints/characters, encoded in UTF-8, or that it can contain at most 256 bytes of UTF-8 encoded text?

**Comment by indutny** [comment]
The intention is to limit number of UTF-8 characters. Would you suggest to
limit the size in bytes instead?

**Comment by DeltaF1** [comment]
The problem is that UTF-8 is a variable-length encoding, so 256 characters could take up between 256 and 1024 bytes. Do you have a specific technical reason to limit it to 256 characters?

**Comment by indutny** [comment]
I agree. The actual space requirements vary a lot for the UTF string of
fixed length.

My intention was to impose the limits on all variable length fields in all
packets that arrive over the wire, so that the byte size of each packet
would be bound.

It seems that limiting the byte length of encoded channel name and message
text is much productive in this regard indeed. Would you be interested in
submitting the change to the documentation?

Thank you!

**Comment by indutny** [comment]
Perhaps 1024 bytes for the message and 128 bytes for the channel would be
best.

**Comment by DeltaF1** [comment]
If you don't have a specific technical need for a maximum of 256 bytes (i.e. some sort of block size for an algorithm) then I don't think there's anything wrong with limiting to a number of characters, not bytes.

**Comment by indutny** [comment]
I guess it is best to leave it as it is for now then. Let's keep the issue open however, as this will have to be addressed in the near future.

**Comment by Qix-** [comment]
I see a reason to limit _bytes_. It makes implementations less prone to attacks where long UTF-8 encoded codepoints could cause overflow errors in weak implementations. Limiting bytes significantly reduces the complexity required to validate a message.

The maximum number of bytes in a UTF-8 codepoint is 8 (though in reality, only up to 4 are used right now) which means a worst-case maximum "valid" length of 127 characters if you limit to 1024 bytes and use null termination, and 255 characters in a "real world" worst case. Add one if you use another field for length.

This also prevents overflow errors by capping maximum leaked overflow to 1024 in poor implementations (even though good protobuf clients should be preventing this from happening anyway).

I would change/phrase it like:

> NOTE: Reason strings MUST be at most 1024 bytes of valid UTF-8 codepoints.

Or since you've already specified 1024 bytes, you could still use "characters" if you so chose - I don't think it would hurt the communication at all.
